packages/material-ui/src/styles/withStyles.js

```diff
@@ -210,6 +213,62 @@ const withStyles = (stylesOrCreator, options = {}) => Component => {
       }
     }
 
+    getClasses() {
+      let generate = false;
+
+      if (!this.disableStylesGeneration) {
+        const sheetManager = this.sheetsManager.get(this.stylesCreatorSaved);
+        const sheetsManagerTheme = sheetManager.get(this.theme);
+        if (sheetsManagerTheme.sheet.classes !== this.lastRenderedClasses) {
+          this.lastRenderedClasses = sheetsManagerTheme.sheet.classes;
+          generate = true;
+        }
+      }
```

Comment:
Same logic as was previously in the `render` method, except now it caches the last rendered classes pointer and toggles the generation on if the classes changed.